### DISCUSSION TITLE
[Chapter 07] 쓰레드 로컬

### DISCUSSION THREAD

**Discussion by BvrPark** [cowaine/spring-boot-for-msa#67] [Chapter 07. Spring AOP, Testing, and Auto Confiugration]
### 책 내용
1. P.402에서 
`어드바이스를 작성할 때 조인 포인트에서 발생하는 예외를 어드바이스 내부에서 직접 처리하는 것은 바람직하지 않다.`
라는 말의 예시에서 트랜잭션 매니저라는 개념이 나왔습니다.
<img src="https://github.com/cowaine/spring-boot-for-msa/assets/84119178/95d689de-b81b-4339-9f8f-02e96e5e263b?raw=true" width="700px">

스프링에서 트랜잭션 매니저가 위의 그림과 같은 개념으로 사용되는 것 같았고 
저 그림에 있는 동기화 매니저는 이 쓰레드 로컬을 사용하여 커넥션을 저장하기 때문에 커넥션을 동기화한다고 하는데
쓰레드 로컬의 개념을 아직 명확하게 이해가 안가서 혹시 아시는 예시가 있으시면 들어주시면 감사하겠습니다.

2. 그리고 저 그림을 그리다가 생각이 든건데 일반적으로 트랜잭션을 시작하고 나서 오류가 생겨서 롤백을 하고 
롤백 작업이 끝난 후에 유지하던 커넥션을 pool로 반환을 하나요? 아니면 그대로 유지를 하나요?

3. 만약 반환을 한다면 롤백을 하던 도중 커넥션을 반환해도 뭔가 커밋을 한 상태가 아니기 때문에 결과적으로 롤백이 될 것인데 
그렇다면 롤백을 할 상황이 되었을 때, 바로 커넥션을 pool로 반환을 하는게 더 빠른게 아닌가요?

</br></br>

### 별도
제가 코드를 따라 치다보니 스프링 빈의 이름이 중복되어서 종종 오류가 걸리는 일이 발생하곤 합니다.
예를 들어, chapter05의 hotelcontroller와 chapter07의 hotelcontroller의 스프링 빈 이름이 같아서 생기는 문제들이요.

저는 이럴 때, 이미 학습한 내용의 챕터는 따로 별칭을 붙여서 해결하는데 다른 분들은 혹시 어떻게 하시는지 궁금해서 질문드립니다.

**Comment by corock** [comment]
1. 예를 들어 현재 회원의 UUID 등 스레드 간 공유할 필요가 있는 값에 대해서 `ThreadLocal` 에 넣고 필요시에 `get()` 으로 불러와서 사용합니다.
2. 커넥션 풀 개수를 지정한 값에 따라 달라지지 않을까요? 이건 직접 확인해 봐야 할 것 같은데, 아시는 분은 댓글 부탁드립니다.
3. 이론상으로는 바로 반환해도 될 것 같은데... 롤백 프로세스에 걸려 있는 배치 이벤트 등이 걸려 있고, 커넥션을 통해 값을 사용할 때 문제될 것 같습니다.
  (이 질문도 2번과 같이 어렵네요...)

### 별도

- `@Qualifier` 애너테이션을 이용하거나 `spring.main.allow-bean-definition-overriding=true` 옵션을 이용하면 될 것 같습니다.

**Reply by corock** [reply]
해당 댓글은 저도 다시 찾아보고 추가로 대댓글 첨부하겠습니다~!

**Comment by Irisation23** [comment]
### 금주 CheckPoint

- JPA 학습 이후, `Transaction Manager` 을 다루고 난 뒤에 다시 돌아 올 것
